### PR TITLE
CLOWDER and tests

### PR DESCRIPTION
**Pull request by Emily-ejag**
The next steps are:

- [x] [@Emily Arteaga](https://yeatman-lab.slack.com/team/U0668LET53P) and [@Anya Ma](https://yeatman-lab.slack.com/team/U01UP86J2MR) will meet to add a fixed order method to jsCat
- [x] [@Emily Arteaga](https://yeatman-lab.slack.com/team/U0668LET53P) will create a new typescript class to manage next item selection in ROAR apps. My strong recommendation is to call this class Clowder because part of it’s job will be to manage multiple Cat instances and “clowder” is the collective noun for cats.
- [x] The Clowder class will take input parameters that define multiple corpora, stimulus presentation rules, etc. and have a method called getNextStimulus or something like that.
- [x] [@Emily Arteaga](https://yeatman-lab.slack.com/team/U0668LET53P) will pilot this new class in ROAR-Letter. The next target after that will be Palabra.

### CODE REVIEW

**Comment by AnyaWMa** [comment]
can you describe what's the purpose of this PR?

File: src/clowder.ts [hunk 1]
```diff
@@ -252,29 +251,27 @@ export class Clowder {
...
-        return allCats.includes(catName);
-      });
-
+      const itemsAndAnswersForCat = itemsAndAnswers.filter(([stim]) =>
+        stim.zetas.some((zeta: ZetaCatMap) => zeta.cats.includes(catName)),
+      );
+      // We are dealing with a single item in this function.  This single item
+      // has an array of zeta parameters for a bunch of different Cats.  We
+      // need to determine if `catName` is present in that list.  So we first
+      // reduce the zetas to get all of the applicabe cat names.
```
Comment (richford) [inline src/clowder.ts:260]: That's a cleaner way to do it! I'd suggest moving this comment back inside of the filter callback and removing the text that says we "first reduce the zeta" and instead saying that we use the `Array.some` method.

File: src/clowder.ts [hunk 2]
```diff
@@ -252,29 +251,27 @@ export class Clowder {
...
+          return [zetaForCat!.zeta, _answer]; // Optional chaining in case zetaForCat is undefined
+        })
+        .filter(([zeta]) => zeta !== undefined); // Filter out undefined zeta values
+
+      // Unzip the zetas and answers, making sure the zetas array contains only Zeta types
+      const [zetas, answers] = _unzip(zetasAndAnswersForCat) as [Zeta[], (0 | 1)[]];
+
+      // Now, pass the filtered zetas and answers to the cat's updateAbilityEstimate method
       this.cats[catName].updateAbilityEstimate(zetas, answers, method);
     }
```
Comment (richford) [inline src/clowder.ts:277]: Can we remove the next TODO comment if you are satisfied that it works?

File: src/clowder.ts [hunk 3]
```diff
@@ -29,6 +29,10 @@ export interface ClowderInput {
    * An optional EarlyStopping instance to use for early stopping.
    */
   earlyStopping?: EarlyStopping;
+  /**
+   * An optional number of items required for each Cat to be considered for early stopping.
+   */
+  numItemsRequired?: CatMap<number>;
```
[the remark this answers is not in the document]
Comment (Emily-ejag) [reply src/clowder.ts:35]: true

File: src/clowder.ts [hunk 4]
```diff
@@ -46,6 +50,7 @@ export class Clowder {
   private _seenItems: Stimulus[];
   private _earlyStopping?: EarlyStopping;
   private readonly _rng: ReturnType<seedrandom>;
+  _numItemsRequired?: CatMap<number>;
```
[the remark this answers is not in the document]
Comment (Emily-ejag) [reply src/clowder.ts:53]: true

File: src/clowder.ts [hunk 5]
```diff
@@ -56,18 +61,15 @@ export class Clowder {
    *
    * @throws {Error} - Throws an error if any item in the corpus has duplicated IRT parameters for any Cat name.
    */
-  constructor({ cats, corpus, randomSeed = null, earlyStopping }: ClowderInput) {
-    // TODO: Need to pass in numItemsRequired so that we know when to stop
-    // providing new items. This may depend on the cat name. For instance,
-    // perhaps numItemsRequired should be an object with cat names as keys and
-    // numItemsRequired as values.
+  constructor({ cats, corpus, randomSeed = null, earlyStopping, numItemsRequired = {} }: ClowderInput) {
```
[the remark this answers is not in the document]
Comment (Emily-ejag) [reply src/clowder.ts:64]: yes

File: src/clowder.ts [hunk 6]
```diff
@@ -212,6 +212,7 @@ export class Clowder {
     method,
     itemSelect,
     randomlySelectUnvalidated = false,
+    returnUndefinedOnExhaustion = false, // New parameter
```
Comment (richford) [inline src/clowder.ts:215]: Can we make the default value for this parameter `true`. I know it will be a hassle in terms of rewriting some of the tests. But I think that Anya wanted the default behavior to be `true`. And I think it makes more sense that way as well.

File: src/clowder.ts [hunk 7]
```diff
@@ -310,7 +312,6 @@ export class Clowder {
         return unvalidatedRemainingItems[randInt];
```
Comment (richford) [inline src/clowder.ts:312]: We should handle the `returnUndefinedOnExhaustion = false` case here as well. If the user has `catToSelect = 'unvalidated'` and `returnUndefinedOnExhaustion = false`, then when `unvalidatedRemainingItems.length === 0`, instead of returning `undefined`, we should return an item that does have parameters for some cat.

Perhaps we can move the
```
const { available, missing } = filterItemsByCatParameterAvailability(this._remainingItems, catToSelect);
```

line above the `if (catToSelect === 'unvalidated') {` line and then draw an item from `missing` in this case.

**Review by richford — CHANGES_REQUESTED** [review]

File: src/clowder.ts (hunk 6 above)
Comment (Emily-ejag) [reply src/clowder.ts:215]: Sure! yes, make sense

File: src/clowder.ts (hunk 7 above)
Comment (Emily-ejag) [reply src/clowder.ts:321]: done!

File: src/__tests__/clowder.test.ts [hunk 8]
```diff
@@ -212,27 +242,30 @@ describe('Clowder Class', () => {
...
         cat1: { method: 'MLE', theta: 0.5 },
+        cat2: { method: 'EAP', theta: -1.0 },
       },
       corpus: [
-        createMultiZetaStimulus('0', [createZetaCatMap([])]),
-        createMultiZetaStimulus('1', [createZetaCatMap(['cat1'])]),
-        createMultiZetaStimulus('2', [createZetaCatMap([])]),
+        createMultiZetaStimulus('0', [createZetaCatMap(['cat2'])]), // Validated for cat2
+        createMultiZetaStimulus('1', [createZetaCatMap(['cat2'])]), // Validated for cat2
+        createMultiZetaStimulus('2', [createZetaCatMap(['cat2'])]), // Validated for cat2
```
Comment (richford) [inline src/__tests__/clowder.test.ts:254]: Can we make one of these items unvalidated? I want to make sure that the unvalidated items wind up in `missing` also.

```suggestion
        createMultiZetaStimulus('1', [createZetaCatMap(['cat2'])]), // Validated for cat1
        createMultiZetaStimulus('2', [createZetaCatMap([])]), // Unvalidated
```

File: src/__tests__/clowder.test.ts [hunk 9]
```diff
@@ -212,27 +242,30 @@ describe('Clowder Class', () => {
...
-      catsToUpdate: ['cat1'],
-      items: [clowder.corpus[1]],
-      answers: [1],
+      returnUndefinedOnExhaustion: false, // Ensure fallback is enabled
     });
+
+    // Should return an item from `missing`, which are items validated for cat2
     expect(nextItem).toBeDefined();
-    expect(['0', '2']).toContain(nextItem?.id);
+    expect(['0', '1', '2']).toContain(nextItem?.id); // Item ID should match any of the items for cat2
```
Comment (richford) [inline src/__tests__/clowder.test.ts:268]: And then here, since we're expecting random behavior, can you do several iterations of this `expect`. Like

```suggestion
    const nDraws = 50;
    for (const _ of Array(nDraws).fill(0)) {
      expect(['0', '1', '2']).toContain(nextItem?.id); // Item ID should match any of the items for cat2
    }
```

File: src/__tests__/clowder.test.ts [hunk 10]
```diff
@@ -125,6 +125,36 @@ describe('Clowder Class', () => {
...
   });
 
+  it('should return an item from missing if catToSelect is "unvalidated", no unvalidated items remain, and returnUndefinedOnExhaustion is false', () => {
+    const clowderInput: ClowderInput = {
+      cats: {
+        cat1: { method: 'MLE', theta: 0.5 },
+      },
+      corpus: [
+        createMultiZetaStimulus('0', [createZetaCatMap(['cat1'])]), // Validated item
+        createMultiZetaStimulus('1', [createZetaCatMap([])]), // Unvalidated item
```
Comment (richford) [inline src/__tests__/clowder.test.ts:135]: Since we're testing a random draw from `missing`, can we include multiple options and then test a few times.

```suggestion
        createMultiZetaStimulus('0', [createZetaCatMap(['cat1'])]), // Validated item
        createMultiZetaStimulus('1', [createZetaCatMap([])]), // Unvalidated item
        createMultiZetaStimulus('2', [createZetaCatMap([])]), // Unvalidated item
```

File: src/__tests__/clowder.test.ts [hunk 11]
```diff
@@ -125,6 +125,36 @@ describe('Clowder Class', () => {
...
+    });
+
+    // Attempt to get another unvalidated item with returnUndefinedOnExhaustion set to false
+    const nextItem = clowder.updateCatAndGetNextItem({
+      catToSelect: 'unvalidated',
+      returnUndefinedOnExhaustion: false,
+    });
+
+    // Should return the validated item since no unvalidated items remain
+    expect(nextItem?.id).toBe('0');
```
Comment (richford) [inline src/__tests__/clowder.test.ts:155]:
```suggestion
    const nDraws = 50;
    for (const _ of Array(nDraws).fill(0)) {
      expect(nextItem?.id).toBe('0');
    }
```

File: src/clowder.ts [hunk 12]
```diff
@@ -212,7 +212,7 @@ export class Clowder {
     method,
     itemSelect,
     randomlySelectUnvalidated = false,
-    returnUndefinedOnExhaustion = false, // New parameter
+    returnUndefinedOnExhaustion = true, // New parameter
```
Comment (richford) [inline src/clowder.ts:215]:
```suggestion
    returnUndefinedOnExhaustion = true,
```

**Review by richford — CHANGES_REQUESTED** [review]

**Review by richford — APPROVED** [review]
Nevermind. I implemented my own comments since I was in there changing other things anyway.

**Comment by richford** [comment]
The test coverage indicates that there are some branches that are not tested in stopping.ts. But I don't think that's a blocker for this PR. I think it's ready for review by @AnyaWMa  and to start implementing in ROAR-Letter.

File: src/__tests__/clowder.test.ts [hunk 13]
```diff
@@ -0,0 +1,585 @@
...
+    expect(nextItem).toBeUndefined();
+  });
+
+  it('can receive one item and answer as an input', () => {
+    const nextItem = clowder.updateCatAndGetNextItem({
+      catToSelect: 'cat1',
+      items: clowder.corpus[0],
+      answers: 1,
+    });
+    expect(nextItem).toBeDefined();
```
Comment (AnyaWMa) [inline src/__tests__/clowder.test.ts:441]: what does "toBeDefined()" mean?

File: src/__tests__/stopping.test.ts [hunk 14]
```diff
@@ -0,0 +1,739 @@
+import { Cat } from '..';
+import { CatMap } from '../type';
+import {
+  EarlyStopping,
+  StopAfterNItems,
+  StopAfterNItemsInput,
+  StopIfSEMeasurementBelowThreshold,
+  StopIfSEMeasurementBelowThresholdInput,
+  StopOnSEMeasurementPlateau,
+  StopOnSEMeasurementPlateauInput,
```
Comment (AnyaWMa) [inline src/__tests__/stopping.test.ts:10]: what is StopOnSEMeasurementPlateau?

File: src/__tests__/stopping.test.ts [hunk 15]
```diff
@@ -0,0 +1,739 @@
+import { Cat } from '..';
```
Comment (AnyaWMa) [inline src/__tests__/stopping.test.ts:1]: do we have tests the cat will stop when item bank is used up?

**Review by AnyaWMa — CHANGES_REQUESTED** [review]
hi I read through a few places, and added some questions. Many of them are probably due to my ignorance, but i just need to know more details how it works. Thank you!

File: src/corpus.ts [hunk 16]
```diff
@@ -0,0 +1,303 @@
...
+ *
+ * This function iterates through each stimulus, extracts parameters relevant to the specified
+ * CAT instances, converts them to the desired format, and returns a cleaned structure of stimuli
+ * with the associated zeta values.
+ */
+export const prepareClowderCorpus = (
+  items: Stimulus[],
+  catNames: string[],
+  delimiter: '.' | string,
+  itemParameterFormat: 'symbolic' | 'semantic' = 'symbolic',
```
[the remark this answers is not in the document]
Comment (Emily-ejag) [reply src/corpus.ts:270]: When using the `prepareClowderCorpus` function, the parameters need to be consistent—either all symbolic (a, b, c, d) or all semantic (discrimination, difficulty, guessing, slipping). We can't mix and match between symbolic and semantic formats in the same input. The function uses the `itemParameterFormat` option to convert everything to the desired format, but we need to pass one type for the input keys.

File: src/stopping.ts [hunk 17]
```diff
@@ -0,0 +1,245 @@
+import { Cat } from './cat';
+import { CatMap } from './type';
+import _uniq from 'lodash/uniq';
+
+/**
+ * Interface for input parameters to EarlyStopping classes.
+ */
+export interface EarlyStoppingInput {
```
[the remark this answers is not in the document]
Comment (Emily-ejag) [reply src/stopping.ts:8]: There are a few combinations we can use for early stopping:

**Logical Operations:** We can choose between and, or, and only for combining multiple stopping criteria:

- **and**: All conditions need to be met to trigger stopping.
- **or**: Any one condition being met will trigger stopping.
- **only**: Only a specific condition is considered (requires you to specify the cat to evaluate).

**Stopping Criteria Classes:**

- **StopAfterNItems**: Stops after a specified number of items.
- **StopOnSEMeasurementPlateau**: Stops if the standard error (SE) of measurement remains stable (within a tolerance) for a specified number of items.
- **StopIfSEMeasurementBelowThreshold**: Stops if the SE measurement drops below a set threshold.

I added these on the readme file :)

File: src/stopping.ts [hunk 18]
```diff
@@ -0,0 +1,245 @@
...
+    if (this._logicalOperation === 'and') {
+      this._earlyStop = conditions.every(Boolean);
+    } else {
+      this._earlyStop = conditions.some(Boolean);
+    }
+  }
+}
+
+/**
+ * Class implementing early stopping based on a plateau in standard error of measurement.
```
[the remark this answers is not in the document]
Comment (Emily-ejag) [reply src/stopping.ts:137]: I don't get it, do you mean we have repetitive classes?

File: src/cat.ts [hunk 19]
```diff
@@ -0,0 +1,319 @@
...
+import { Stimulus, Zeta } from './type';
+import { itemResponseFunction, fisherInformation, normal, findClosest } from './utils';
+import { validateZetaParams, fillZetaDefaults } from './corpus';
+import seedrandom from 'seedrandom';
+import _clamp from 'lodash/clamp';
+import _cloneDeep from 'lodash/cloneDeep';
+
+const abilityPrior = normal();
+
+export interface CatInput {
```
[the remark this answers is not in the document]
Comment (Emily-ejag) [reply src/cat.ts:12]: CatInput is just a configuration interface that defines the setup options for creating a Cat, like method or theta. On the other hand, Cat is the actual class that manages the behavior and state.

Basically, CatInput is for setup, and Cat does the work!

File: src/__tests__/stopping.test.ts (hunk 14 above)
Comment (Emily-ejag) [reply src/__tests__/stopping.test.ts:10]: In simpler terms, if the SE doesn't significantly change for a set number of items, the process stops early because it indicates that the ability estimate is no longer improving. It's used to avoid unnecessary trials once the measurement has plateaued.

File: src/__tests__/stopping.test.ts (hunk 15 above)
Comment (Emily-ejag) [reply src/__tests__/stopping.test.ts:1]: Yes, all lines are covered now

**Review by richford — CHANGES_REQUESTED** [review]

File: src/stopping.ts (hunk 18 above)
[the remark this answers is not in the document]
Comment (richford) [reply src/stopping.ts:137]: This one stops if the SEMeasurement plateaus, meaning that is fails to decrease for a certain number of trials (called `patience`). E.g., if you have

| trial number | SEMeasurement |
| :------------ | :---------------- |
| 1 | 1.0 |
| 2 | 0.8 |
| 3 | 0.6 |
| 4 | 0.4 |
| 5 | 0.4 |
| 6 | 0.4 |
| 7 | 0.4 |

If the `patience` was set to 3, then the early stopping would have been triggered on trial 6 because the SE had failed to decrease for 3 consecutive trials.

File: src/__tests__/clowder.test.ts [hunk 20]
```diff
@@ -0,0 +1,585 @@
...
+  zeta,
+});
+
+describe('Clowder Class', () => {
+  let clowder: Clowder;
+
+  beforeEach(() => {
+    const clowderInput: ClowderInput = {
+      cats: {
+        cat1: { method: 'MLE', theta: 0.5 },
```
[the remark this answers is not in the document]
Comment (richford) [reply src/__tests__/clowder.test.ts:30]: In reality, it is a `Cat` instance. But for unit testing, it is enough to provide an object that "looks" like a `Cat` instance.

File: src/__tests__/clowder.test.ts (hunk 13 above)
Comment (richford) [reply src/__tests__/clowder.test.ts:441]: This is a Jest "matcher" used for unit testing https://jestjs.io/docs/expect#tobedefined

File: src/__tests__/stopping.test.ts (hunk 14 above)
Comment (richford) [reply src/__tests__/stopping.test.ts:10]: See my comment here: https://github.com/yeatmanlab/jsCAT/pull/25#discussion_r1813477978

File: src/clowder.ts [hunk 21]
```diff
@@ -257,33 +257,35 @@ export class Clowder {
...
+              return [zetaForCat!.zeta, _answer]; // Optional chaining in case zetaForCat is undefined
+            })
+            .filter(([zeta]) => zeta !== undefined); // Filter out undefined zeta values
+
+          // Unzip the zetas and answers, making sure the zetas array contains only Zeta types
+          const [zetas, answers] = _unzip(zetasAndAnswersForCat) as [Zeta[], (0 | 1)[]];
+
+          // Now call updateAbilityEstimates for this cat
+          this.updateAbilityEstimates([catName], zetas, answers, method);
+        }
```
[the remark this answers is not in the document]
Comment (Emily-ejag) [reply src/clowder.ts:288]: I don't think so, because we have the for loop that goes to all the catsToUpdate, so it ran the practice cats, but we don't need to update those, just the ones like lowerLetter, upperLetter

File: src/stopping.ts [hunk 22]
```diff
@@ -96,15 +96,39 @@ export abstract class EarlyStopping {
...
+    // Iterate over each cat and update the _nItems map
+    for (const catName in cats) {
+      const cat = cats[catName];
+      const nItems = cat.nItems; // Get the current number of items for this cat
+
+      // Update the _nItems map with the current nItems value
+      if (nItems !== undefined) {
+        this._nItems[catName] = nItems; // Make sure nItems is set for this cat
+      }
+    }
```
[the remark this answers is not in the document]
Comment (Emily-ejag) [reply src/stopping.ts]: done

**Review by richford — APPROVED** [review]

File: src/stopping.ts (hunk 18 above)
[the remark this answers is not in the document]
Comment (AnyaWMa) [reply src/stopping.ts:126]: got it! thank you!

File: src/__tests__/stopping.test.ts (hunk 15 above)
Comment (AnyaWMa) [reply src/__tests__/stopping.test.ts:1]: where is the test when item bank is used up?

File: src/corpus.ts (hunk 16 above)
[the remark this answers is not in the document]
Comment (AnyaWMa) [reply src/corpus.ts:270]: is this requirement documented somewhere?

File: src/__tests__/clowder.test.ts (hunk 13 above)
Comment (AnyaWMa) [reply src/__tests__/clowder.test.ts:441]: okay, thanks for explaining.

File: README.md [hunk 23]
```diff
@@ -42,22 +42,56 @@ const stimuli = [{difficulty: -3, item: 'item1'}, {difficulty: -2,  item: 'item2
...
+
+### How Combinations Work
+
+You can mix and match these criteria with different logical operations, giving you a range of configurations for early stopping. For example:
+
+- Using **`and`** with both `StopAfterNItems` and `StopIfSEMeasurementBelowThreshold` means stopping will only occur if both conditions are satisfied.
+- Using **`or`** with `StopOnSEMeasurementPlateau` and `StopAfterNItems` allows early stopping if either condition is met.
+
+If you need more details or a specific example documented, feel free to ask!
+
```
Comment (AnyaWMa) [inline README.md:73]: thanks for the document. can you give sample lines of code here?

File: README.md [hunk 24]
```diff
@@ -42,22 +42,56 @@ const stimuli = [{difficulty: -3, item: 'item1'}, {difficulty: -2,  item: 'item2
 const nextItem = cat.findNextItem(stimuli, 'MFI');
 ```
```
Comment (AnyaWMa) [inline README.md:44]: can you add documentation about the accepted stimuli types?

File: src/cat.ts (hunk 19 above)
[the remark this answers is not in the document]
Comment (AnyaWMa) [reply src/cat.ts:12]: i am confused. does it mean the user need to create CatInput instead of Cat?

if yes, can you add documentation in the README to explain how to use jsClowder 

And will that conflict the use if user only wants to use jsCAT instead of jsClowder?

File: src/corpus.ts (hunk 16 above)
[the remark this answers is not in the document]
Comment (AnyaWMa) [reply src/corpus.ts:270]: i would suggest document it in README with sample code

File: src/stopping.ts (hunk 17 above)
[the remark this answers is not in the document]
Comment (AnyaWMa) [reply src/stopping.ts:8]: i think providing sample code in README would be helpful.

File: README.md (hunk 24 above)
Comment (AnyaWMa) [inline README.md:44]: more broadly, can we document and show sample code: how to set up a jsClowder, and some basic functions to run a clowder.

**Review by AnyaWMa — CHANGES_REQUESTED** [review]
Thank you for lots of improvement! I understand the code better now, but I will still appreciate more detailed documentation in the README. Specficially, I want to know will jsClowder will conflict any use if the user will want to use jsCAT?

File: src/__tests__/stopping.test.ts (hunk 15 above)
Comment (Emily-ejag) [reply src/__tests__/stopping.test.ts:1]: line 303 is the test if you want a set number, if you don't add early stopping it will go trough all

**Review by richford — DISMISSED** [review]

**Review by AnyaWMa — CHANGES_REQUESTED** [review]
sorry for another round of clarification. 

I am asking because I found in my current jsCAT: the desired zeta type is actually a mix of symbolic and semantic (a, difficulty, c, and d.). I am okay to just be all semantic or all symbolic, but I want to make sure all places that are hard-coded in the code will function as expected for this transition. Thanks!

File: src/cat.ts [hunk 25]
```diff
@@ -0,0 +1,319 @@
...
+
+    stimuliAddFisher.sort((a, b) => b.fisherInformation - a.fisherInformation);
+    stimuliAddFisher.forEach((stimulus: Stimulus) => {
+      delete stimulus['fisherInformation'];
+    });
+    return {
+      nextStimulus: stimuliAddFisher[0],
+      remainingStimuli: stimuliAddFisher.slice(1).sort((a: Stimulus, b: Stimulus) => a.difficulty! - b.difficulty!),
+    };
+  }
```
[the remark this answers is not in the document]
Comment (richford) [reply src/cat.ts:255]: The returned stimuli will be semantic, in keeping with the legacy behavior of jsCat.

File: src/cat.ts (hunk 19 above)
[the remark this answers is not in the document]
Comment (richford) [reply src/cat.ts:12]: We clarified this in person I think but just to follow up here. `CatInput` is just an interface defining the input format that `Cat` expects when it is being instantiated. The user creates a new `Cat` instance by passing in parameters that conform to the `CatInput` interface. It's actually already defined in jsCat. Emily didn't add it in this PR. @AnyaWMa , you added it two years ago in [this commit](https://github.com/yeatmanlab/jsCAT/commit/60571e69318c20bbf575a80ba02cefd6b17af55e).

File: src/cat.ts [hunk 26]
```diff
@@ -0,0 +1,319 @@
...
+    arr = arr.map((stim) => fillZetaDefaults(stim, 'semantic'));
+
+    if (this.nItems < this.nStartItems) {
+      selector = this.startSelect;
+    }
+    if (selector !== 'mfi' && selector !== 'fixed') {
+      // for mfi, we sort the arr by fisher information in the private function to select the best item,
+      // and then sort by difficulty to return the remainingStimuli
+      // for fixed, we want to keep the corpus order as input
+      arr.sort((a: Stimulus, b: Stimulus) => a.difficulty! - b.difficulty!);
```
[the remark this answers is not in the document]
Comment (richford) [reply src/cat.ts:223]: Nope. Nine lines above, Emily ensures that the item array is in the semantic format using the `fillZetaDefaults` method.

File: src/cat.ts (hunk 25 above)
[the remark this answers is not in the document]
Comment (richford) [reply src/cat.ts:255]: The returned stimuli will be semantic, in keeping with the legacy behavior of jsCat.

File: src/cat.ts (hunk 26 above)
[the remark this answers is not in the document]
Comment (richford) [reply src/cat.ts:223]: Nope. Nine lines above, Emily ensures that the item array is in the semantic format using the `fillZetaDefaults` method.

File: src/cat.ts (hunk 19 above)
[the remark this answers is not in the document]
Comment (richford) [reply src/cat.ts:12]: I think we clarified this in our Slack huddle but I'll just document the conversation here. `CatInput` is an interface defining the input format that `Cat` expects when it is being instantiated. The user creates a new `Cat` instance by passing in parameters that conform to the `CatInput` interface. It's actually already defined in jsCat. Emily didn't add it in this PR. @AnyaWMa , you added it two years ago in [this commit](https://github.com/yeatmanlab/jsCAT/commit/60571e69318c20bbf575a80ba02cefd6b17af55e).

File: src/__tests__/stopping.test.ts (hunk 15 above)
Comment (richford) [reply src/__tests__/stopping.test.ts:1]: And line 98 in `src/__tests__/clowder.test.ts` tests that Clowder will return undefined when an item bank for a specified `catToSelect` is used up.

File: src/__tests__/cat.test.ts [hunk 27]
```diff
@@ -0,0 +1,233 @@
...
+      received = cat4.findNextItem(stimuliSorted, undefined, deepCopy);
+      expect(received.nextStimulus).toEqual(stimuliSorted[index]);
+
+      for (let i = 0; i < 3; i++) {
+        const remainingStimuli = received.remainingStimuli;
+        index = Math.floor(rng() * remainingStimuli.length);
+        received = cat4.findNextItem(remainingStimuli, undefined, deepCopy);
+        expect(received.nextStimulus).toEqual(remainingStimuli[index]);
+      }
+    });
```
[the remark this answers is not in the document]
Comment (richford) [reply src/__tests__/cat.test.ts:166]: `deepCopy` is part of the original jsCat code. It dictates whether the input array should be deep copied before sorting in the `findNextItem` method. While the `deepCopy` parameter itself isn't new (it's been around since [this commit](https://github.com/yeatmanlab/jsCAT/commit/8c7611325f52c4151943327d480b921f89643538#diff-a2a171449d862fe29692ce031981047d7ab755ae7f84c707aef80701b3ea0c80), Emily noticed that it wasn't being tested, so now the unit tests test whether `findNextItem` works as expected with both `deepCopy=true` and `deepCopy=false`.

**Review by AnyaWMa — DISMISSED** [review]

**Comment by AnyaWMa** [comment]
i approved the PR, but I will mention this version will break current use of jsCAT in apps. The structure of taking zetas is completely different, and the README is not valid too. it is important to let the users know (swr, vocab, comp, and maybe levante). Thank you!

**Review by richford — DISMISSED** [review]

File: README.md [hunk 28]
```diff
@@ -37,27 +45,205 @@ const numItems = cat.nItems;
...
-Lucas Duailibe, irt-js, (2019), GitHub repository, https://github.com/geekie/irt-js
+- Lucas Duailibe, irt-js, (2019), GitHub repository, https://github.com/geekie/irt-js
 
 ## License
+
 jsCAT is distributed under the [ISC license](LICENSE).
+
+
+# Clowder Usage Guide
+
```
Comment (AnyaWMa) [inline README.md:109]: maybe put Clowder Usage Guide before the References

File: README.md [hunk 29]
```diff
@@ -37,27 +45,205 @@ const numItems = cat.nItems;
 
 // find the next available item from an input array of stimuli based on a selection method
 
-const stimuli = [{difficulty: -3, item: 'item1'}, {difficulty: -2,  item: 'item2'}];
+const stimuli = [{  discrimination: 1, difficulty: -2, guessing: 0, slipping: 1, item = "item1" },{ discrimination: 1, difficulty: 3, guessing: 0, slipping: 1, item = "item2" }];
```
Comment (AnyaWMa) [inline README.md:49]: do authors need to specify all the params? (a, b, c, d)?

File: README.md [hunk 30]
```diff
@@ -37,27 +45,205 @@ const numItems = cat.nItems;
 
 // find the next available item from an input array of stimuli based on a selection method
 
-const stimuli = [{difficulty: -3, item: 'item1'}, {difficulty: -2,  item: 'item2'}];
```
Comment (AnyaWMa) [inline README.md:40]: I will make a note about version change to users: "For existing jsCAT users: to make your applications compatible to the updated jsCAT version, you will need to pass the stimuli in the following way:"

File: README.md (hunk 29 above)
Comment (AnyaWMa) [reply README.md:49]: if they have to, what are default values they should put if they only care about difficulty? a = 1, c = 0, d = 1

**Review by AnyaWMa — CHANGES_REQUESTED** [review]
Thank you for putting detailed README! I just have a few things before I can approve.

File: README.md (hunk 30 above)
Comment (Emily-ejag) [reply README.md:40]: Sure! I added it before the code so it is clear

File: README.md (hunk 29 above)
Comment (Emily-ejag) [reply README.md:49]: Good question, yes they have to put all values. Let me ask what would be the default

Comment (richford) [reply README.md:49]: I don't think they need to specify all of the values due to the `fillDefaultZeta` function in `corpus.ts`. But @Emily-ejag, could you do a quick test to verify that it works with missing params?

**Review by richford — DISMISSED** [review]

File: README.md (hunk 29 above)
Comment (Emily-ejag) [reply README.md:49]: We have a function on clowder called `fillZetaDefaults` which will fiil out the values to be `export const defaultZeta = (desiredFormat: 'symbolic' | 'semantic' = 'symbolic'): Zeta => {
  const defaultZeta: Zeta = {
    a: 1,
    b: 0,
    c: 0,
    d: 1,
  };

  return convertZeta(defaultZeta, desiredFormat);
};` I can add the documentation for this

File: README.md (hunk 24 above)
Comment (Emily-ejag) [reply README.md:52]: Done!

File: README.md (hunk 28 above)
Comment (Emily-ejag) [reply README.md:109]: Good idea!

**Review by AnyaWMa — APPROVED** [review]

**Review by richford — APPROVED** [review]
